pallene/translator.lua

```diff
@@ -31,77 +31,124 @@ local Translator = util.Class()
 function Translator:init()
     self.last_index = 1
     self.partials = {}
+    self.exports = {}
+    self.input = false
     return self
 end
 
-function Translator:add_previous(input, stop_index)
-    local partial = input:sub(self.last_index, stop_index)
+function Translator:add_previous(stop_index)
+    local partial = self.input:sub(self.last_index, stop_index)
     table.insert(self.partials, partial)
     self.last_index = stop_index + 1
 end
 
-function Translator:add_whitespace(input, start_index, stop_index)
-    self:add_previous(input, start_index - 1)
+function Translator:add_whitespace(start_index, stop_index)
+    self:add_previous(start_index - 1)
 
-    local region = input:sub(start_index, stop_index)
+    local region = self.input:sub(start_index, stop_index)
     local partial = region:gsub("%S", " ")
     table.insert(self.partials, partial)
 
     self.last_index = stop_index + 1
 end
 
+function Translator:add_local(start_index, stop_index)
+    self:add_previous(start_index - 1)
+    -- The export keyword is six characters long, whereas the local keyword is five characters
+    -- long. Therefore, we pad the keyword with a space. We could add the space to the left, too.
+    -- But it seems more "natural" on the right.
+    table.insert(self.partials, "local ")
+
+    self.last_index = stop_index
+end
+
+function Translator:add_exports()
+    if #self.exports > 0 then
+        table.insert(self.partials, "\nreturn {\n")
+        for _, export in ipairs(self.exports) do
+            local pair = string.format("    %s = %s,\n", export, export)
+            table.insert(self.partials, pair)
+        end
+        table.insert(self.partials, "}\n")
+    end
+end
+
+function Translator:translate_var_decl(decl)
+    if decl.type then
+        -- Remove the colon but retain any adjacent comment to the right.
+        self:add_whitespace(decl.col_loc.pos, decl.col_loc.pos)
+        -- Remove the type annotation but exclude the next token.
+        self:add_whitespace(decl.type.loc.pos, decl.end_loc.pos - 1)
+    end
+end
+
+function Translator:translate_tl_variables(node)
```

Comment:
We could have a single `translate_toplevel` function instead of one function for each kind of toplevel thing.